### PR TITLE
erlang: patch sh path

### PR DESCRIPTION
**Pull request by liamdiprose**
## Description of changes

RabbitMQ 3.13 crashes because it requires Erlang/OTP 26 and hits [this call to `sh`](https://github.com/erlang/otp/blob/maint-26/lib/os_mon/src/disksup.erl#L243). This patches the call to `sh` with its full path, `${stdenv.shell}/bin/sh`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

## Questions for the reviewers

- [x] Uses `substituteInPlace`, would you prefer a patch file instead?
- [ ] This issue only started happening in OTP 26, is generic-builder.nix the best place for it?

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by yurrriq** [comment]
Thanks! Should we use pkgs.sh or whatever instead? Otherwise this lgtm

**Comment by turion** [comment]
@ofborg test rabbitmq

**Comment by liamdiprose** [comment]
I've changed `${stdenv.shell}/bin/sh` to `${bash}/bin/sh`, since it is run on the host and isn't related to the build environment.

File: pkgs/development/interpreters/erlang/generic-builder.nix [hunk 1]
```diff
@@ -113,6 +114,8 @@ stdenv.mkDerivation ({
 
   postPatch = ''
     patchShebangs make
+    substituteInPlace lib/os_mon/src/disksup.erl \
+      --replace '"sh ' '"${bash}/bin/sh '
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/generic-builder.nix:118]:
```suggestion
    substituteInPlace lib/os_mon/src/disksup.erl \
      --replace '"sh ' '"${bash}/bin/sh '
```
Currently `substituteInPlace` has `--replace` deprecated.
Suggested usage: `--replace-fail`, `--replace-warn` and `--replace-quiet`

**Comment by superherointj** [comment]
> I've changed `${stdenv.shell}/bin/sh` to `${bash}/bin/sh`, since it is run on the host and isn't related to the build environment.

When I attempted to use NixNG (which uses nixpkgs) [a Nix distribution for containers, like alpine], one important goal was to replace `bash` with `busybox ash` because it is much smaller. I think, using it with default shell makes sense. But I'm also not certain of the nuanced consequences of this. I hear you want to think of the `host` but I saw this as the build of the `erlang` package.

Why is it a problem to keep `${stdenv.shell}/bin/sh`? Wouldn't it offer better compatibility for other shells like `ash`?

**Comment by superherointj** [comment]
>  Uses substituteInPlace, would you prefer a patch file instead?

No. The usage of a patch file should be avoided for simple things like this. As long as the text matching is done properly.

File: pkgs/development/interpreters/erlang/generic-builder.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/development/interpreters/erlang/generic-builder.nix:118]: > This issue only started happening in OTP 26, is generic-builder.nix the best place for it?

Considering this, will it fail for other Erlang versions? Humm...

**Comment by liamdiprose** [comment]
> Why is it a problem to keep ${stdenv.shell}/bin/sh? Wouldn't it offer better compatibility for other shells like ash?

I've worked on cross compiling where the `buildInputs` and `nativeBuildInputs` are separate from each other. The `sh` will be ran on the host system, so it doesn't make sense to use the `sh` used to build the package.

The original commit patched `/bin/sh` because I thought it would be a tiny efficient blob of bytes, but its actually a symlink to `${bash}/bin/sh`. I would have liked to see ash there as well.

Since bash is probably already loaded, maybe it's still efficient?

**Comment by superherointj** [comment]
I suspect `erlang/28.nix` will need same patch. I think would be best to assume all forward versions will need this patch.
And then, making it conditional for previous versions at `generic.nix` might be best.

**Comment by superherointj** [comment]
> Since bash is probably already loaded, maybe it's still efficient?

As it is runtime, I think we should use: "${runtimeShell}"

File: pkgs/development/interpreters/erlang/generic-builder.nix [hunk 2]
```diff
@@ -47,6 +48,7 @@
 , odbcPackages ? [ unixODBC ]
 , opensslPackage ? openssl
 , wxPackages ? [ libGL libGLU wxGTK xorg.libX11 wrapGAppsHook3 ]
+, patchShell ? true
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/generic-builder.nix:51]:
```suggestion
, patchShell ? true
```
Sorry, I did not mean this.
I don't think the user needs to tweak this.
I meant just to check for versions directly.

File: pkgs/development/interpreters/erlang/24.nix [hunk 3]
```diff
@@ -3,4 +3,5 @@
 mkDerivation {
   version = "24.3.4.17";
   sha256 = "sha256-V26pZEyFo+c+ztDDkjDNFK6LTw5xzF8gQYepWGNlGKg=";
+  patchShell = false;
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/24.nix:6]:
```suggestion
```

File: pkgs/development/interpreters/erlang/25.nix [hunk 4]
```diff
@@ -3,4 +3,5 @@
 mkDerivation {
   version = "25.3.2.12";
   sha256 = "sha256-ZqyJtKr0QdzuuzKnJ/S6Z4cGLWolVMVWcMuAHyOGzkQ=";
+  patchShell = false;
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/25.nix:6]:
```suggestion
```

File: pkgs/development/interpreters/erlang/generic-builder.nix [hunk 5]
```diff
@@ -11,6 +11,7 @@
 , ncurses
 , openssl
 , perl
+, bash
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/generic-builder.nix:14]:
```suggestion
```

File: pkgs/development/interpreters/erlang/generic-builder.nix (hunk 2 above)
Comment (liamdiprose) [reply pkgs/development/interpreters/erlang/generic-builder.nix:51]: I think the only users of those options are the sibling `24.nix`, `25.nix` files?

I can't see a good way to check the versions directly, other than using the major-version nix files themselves.

i.e. `25.nix` specifies `version = "25.3.2.12";`.

File: pkgs/development/interpreters/erlang/generic-builder.nix [hunk 6]
```diff
@@ -115,6 +117,9 @@ stdenv.mkDerivation ({
     patchShebangs make
 
     ${postPatch}
+  '' + optionalString patchShell ''
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/generic-builder.nix:120]:
```suggestion
  '' + optionalString patchShell ''
```
I think you should test version here replacing `patchShell`. using `lib.versionOlder`. Does it make sense?

File: pkgs/development/interpreters/erlang/generic-builder.nix [hunk 7]
```diff
@@ -115,6 +117,9 @@ stdenv.mkDerivation ({
...
+    substituteInPlace lib/os_mon/src/disksup.erl \
+      --replace-fail '"sh ' '"${bash}/bin/sh '
```
Comment (superherointj) [inline pkgs/development/interpreters/erlang/generic-builder.nix:122]:
```suggestion
      --replace-fail '"sh ' '"${runtimeShell} '
```
(Remember to add `runtimeShell` at top arguments too.)

**Comment by liamdiprose** [comment]
Thanks for the help, I've made the changes.

**Comment by superherointj** [comment]
@ofborg test rabbitmq
@ofborg build pkgsMusl.erlang

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 317646` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>37 packages built:</summary>
  <ul>
    <li>akkoma</li>
    <li>cl</li>
    <li>couchdb3</li>
    <li>elixir</li>
    <li>elixir-ls</li>
    <li>elixir_1_10</li>
    <li>elixir_1_11</li>
    <li>elixir_1_12</li>
    <li>elixir_1_13</li>
    <li>elixir_1_14</li>
    <li>elixir_1_15</li>
    <li>elvis-erlang</li>
    <li>erlang</li>
    <li>erlang-ls</li>
    <li>erlang_26</li>
    <li>erlang_27</li>
    <li>erlang_nox</li>
    <li>erlfmt</li>
    <li>ex_doc</li>
    <li>lexical</li>
    <li>lfe</li>
    <li>livebook</li>
    <li>mercury</li>
    <li>mix2nix</li>
    <li>mobilizon</li>
    <li>notmuch-bower</li>
    <li>plausible</li>
    <li>pleroma</li>
    <li>rabbitmq-server</li>
    <li>rabbitmq-server.doc</li>
    <li>rabbitmq-server.man</li>
    <li>rebar</li>
    <li>rebar3</li>
    <li>sonic-pi</li>
    <li>tsung</li>
    <li>wings</li>
    <li>yaws</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Pull request merged**